keps/sig-cluster-lifecycle/kubeadm-config-ha-support.md

```diff
@@ -0,0 +1,161 @@
+---
+kep-number: TBD
+title: Augment Kubeadm Config to Enable Upgrades of HA Clusters
+authors:
+  - "@mattkelly"
+owning-sig: sig-cluster-lifecycle
+reviewers:
+  - "@timothysc"
+approvers:
+  - TBD
+editor:
+  - "@mattkelly"
+creation-date: 2018-03-08
+last-updated: 2018-03-19
+status: provisional
+see-also:
+  - [KEP kubeadm join --master workflow](https://github.com/kubernetes/community/pull/1707) (in progress)
+  - [Upgrading kubeadm HA clusters from 1.9.x to 1.9.y](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm-upgrade-ha)
+---
+
+# Augment Kubeadm Config to Enable Upgrades of HA Clusters
+
+## Table of Contents
+
+* [Augment Kubeadm Config to Enable Upgrades of HA Clusters](#augment-kubeadm-config-to-enable-upgrades-of-ha-clusters)
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+   * [Goals](#goals)
+   * [Non-Goals](#non-goals)
+   * [Challenges and Open Questions](#challenges-and-open-questions)
+* [Proposal](#proposal)
+   * [Implementation Details](#implementation-details)
+      * [Background](#background)
+      * [Adding Additional Master-Specific ConfigMaps](#adding-additional-master-specific-configmaps)
+         * [Key Design Considerations and Benefits](#key-design-considerations-and-benefits)
+            * [Parallel Node Creation](#parallel-node-creation)
+            * [Guaranteed Consistent kubeadm-config](#guaranteed-consistent-kubeadm-config)
+   * [Risks and Mitigations](#risks-and-mitigations)
+      * [Migrating Existing Clusters](#migrating-existing-clusters)
+      * [More Complex User Experience for Overriding Configuration](#more-complex-user-experience-for-overriding-configuration)
+* [Implementation History](#implementation-history)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+
+## Summary
+
+One of the first steps of the upgrade process is to retrieve the `kubeadm-config` ConfigMap that is stored at `kubeadm init` time in order to get any relevant cluster configuration.
+The current `kubeadm-config` ConfigMap was designed to support upgrades of clusters with a single master.
+As we move towards supporting High Availability (HA) natively in kubeadm, the persistent configuration must be enhanced in order to store information unique to each master node.
+In particular, master node-specific information is required in order for kubeadm to identify which control plane static pods belong to the current node during an upgrade (the static pods are identifiable by the `nodeName` field embedded in the pod name).
```

Comment:
Why is this specific to static pods?

Reply:
It may not be - I'll make it more clear that this is just an example of how the node-specific information is used.